File: spring-core/src/main/java/com/example/spring/core/api/dockerhub/DockerHubController.java
package com.example.spring.core.api.dockerhub;

import com.example.spring.core.api.dockerhub.DockerHubService.ImageTag;
import java.util.List;
import lombok.RequiredArgsConstructor;
import lombok.extern.log4j.Log4j2;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import reactor.core.publisher.Mono;

@RestController
@RequestMapping("/dockerhub")
@RequiredArgsConstructor
@Log4j2
public class DockerHubController {

  private final DockerHubService service;

  @GetMapping("/images/{namespace}/{repository}/tags")
  public Mono<List<ImageTag>> getAllTags(
      @PathVariable String namespace,
      @PathVariable String repository
  ) {
    return service.listImageTags(namespace, repository);
  }

  @GetMapping("/images/{namespace}/{repository}/tags/{tag}")
  public Mono<ImageTag> getTag(
      @PathVariable String namespace,
      @PathVariable String repository,
      @PathVariable String tag
  ) {
    return service.getImageTag(namespace, repository, tag);
  }

  @GetMapping("/images/{namespace}/{repository}/tags/matching")
  public Mono<List<ImageTag>> getMatchingTags(
      @PathVariable String namespace,
      @PathVariable String repository,
      @RequestParam String digest
  ) {
    return service.listImageTags(namespace, repository)
        .map(imageTags -> service.findMatchingTags(imageTags, digest));
  }

}

File: tools/src/main/kotlin/com/example/spring/tools/dockerhub/DockerHubController.kt
package com.example.spring.tools.dockerhub

import com.example.spring.tools.dockerhub.DockerHubService.ImageTag
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/dockerhub")
class DockerHubController(private val service: DockerHubService) {

  @GetMapping("/images/{namespace}/{repository}/tags")
  suspend fun getAllTags(
    @PathVariable namespace: String,
    @PathVariable repository: String
  ): List<ImageTag> {
    return service.listImageTags(namespace, repository)
  }

  @GetMapping("/images/{namespace}/{repository}/tags/{tag}")
  suspend fun getTag(
    @PathVariable namespace: String,
    @PathVariable repository: String,
    @PathVariable tag: String
  ): ImageTag {
    return service.getImageTag(namespace, repository, tag);
  }

  @GetMapping("/images/{namespace}/{repository}/tags/matching")
  suspend fun getMatchingTags(
    @PathVariable namespace: String,
    @PathVariable repository: String,
    @RequestParam digest: String
  ): List<ImageTag> {
    val imageTags = service.listImageTags(namespace, repository)
    return service.findMatchingTags(imageTags, digest)
  }
}